abci/check_tx.go

```diff
@@ -0,0 +1,186 @@
+		if err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("failed to decode tx: %w", err), 0, 0, nil, false)
+		}
+
+		// Attempt to get the bid info of the transaction.
+		bidInfo, err := handler.mempool.GetAuctionBidInfo(sdkTx)
+		if err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("failed to get auction bid info: %w", err), 0, 0, nil, false)
+		}
+
+		// If this is not a bid transaction, we just execute it normally.
+		if bidInfo == nil {
+			return handler.baseApp.CheckTx(req)
+		}
+
+		// We attempt to get the latest committed state in order to verify transactions
+		// as if they were to be executed at the top of the block. After verification, this
+		// context will be discarded and will not apply any state changes.
+		ctx := handler.GetContextForBidTx(req)
+
+		// Verify the bid transaction.
+		ctx, err = handler.anteHandler(ctx, sdkTx, false)
+		if err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("invalid bid tx; failed to execute ante handler: %w", err), 0, 0, nil, false)
+		}
+
+		// Verify all of the bundled transactions.
+		for _, tx := range bidInfo.Transactions {
+			bundledTx, err := handler.mempool.WrapBundleTransaction(tx)
+			if err != nil {
+				return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("invalid bid tx; failed to decode bundled tx: %w", err), 0, 0, nil, false)
+			}
+
+			bidInfo, err := handler.mempool.GetAuctionBidInfo(bundledTx)
+			if err != nil {
+				return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("invalid bid tx; failed to get auction bid info: %w", err), 0, 0, nil, false)
+			}
+
+			// Bid txs cannot be included in bundled txs.
+			if bidInfo != nil {
+				return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("invalid bid tx; bundled tx cannot be a bid tx"), 0, 0, nil, false)
+			}
+
+			if ctx, err = handler.anteHandler(ctx, bundledTx, false); err != nil {
+				return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("invalid bid tx; failed to execute bundled transaction: %w", err), 0, 0, nil, false)
+			}
+		}
+
+		// If the bid transaction is valid, we know we can insert it into the mempool for consideration in the next block.
+		if err := handler.mempool.Insert(ctx, sdkTx); err != nil {
+			return sdkerrors.ResponseCheckTxWithEvents(fmt.Errorf("invalid bid tx; failed to insert bid transaction into mempool: %w", err), 0, 0, nil, false)
+		}
+
+		return cometabci.ResponseCheckTx{Code: cometabci.CodeTypeOK}
+	}
+}
+
+// GetContextForBidTx is returns the latest committed state and sets the context given
+// the checkTx request.
+func (handler *CheckTxHandler) GetContextForBidTx(req cometabci.RequestCheckTx) sdk.Context {
+	// Retrieve the commit multi-store which is used to retrieve the latest committed state.
+	ms := handler.baseApp.CommitMultiStore().CacheMultiStore()
+
+	// Create a new context based off of the latest committed state.
+	ctx, _ := sdk.NewContext(ms, tmproto.Header{}, false, handler.baseApp.Logger()).CacheContext()
```

Comment:
Don't we want to populate the header with relevant fields? At the very least the height. That way we can remove the  `WithBlockHeight` call below.

Reply:
bump :)